In [1]:
import pandas as pd

In [3]:
Data = pd.read_csv('Resources/mainDS.csv')

1.Display top 5 rows of the data

In [4]:
Data.head()

ID  Loan Amount  Funded Amount  Funded Amount Investor  Term  \
0  65087372        10000          32236             12329.36286    59   
1   1450153         3609          11940             12191.99692    59   
2   1969101        28276           9311             21603.22455    59   
3   6651430        11170           6954             17877.15585    59   
4  14354669        16890          13226             13539.92667    59   

  Batch Enrolled  Interest Rate Grade Sub Grade Home Ownership  ...  \
0     BAT2522922      11.135007     B        C4       MORTGAGE  ...   
1     BAT1586599      12.237563     C        D3           RENT  ...   
2     BAT2136391      12.545884     F        D4       MORTGAGE  ...   
3     BAT2428731      16.731201     C        C3       MORTGAGE  ...   
4     BAT5341619      15.008300     C        D4       MORTGAGE  ...   

    Recoveries Collection Recovery Fee Collection 12 months Medical  \
0     2.498291                0.793724                            0   
1     2.377215                0.974821                            0   
2     4.316277                1.020075                            0   
3     0.107020                0.749971                            0   
4  1294.818751                0.368953                            0   

  Application Type  Last week Pay  Accounts Delinquent  \
0       INDIVIDUAL             49                    0   
1       INDIVIDUAL            109                    0   
2       INDIVIDUAL             66                    0   
3       INDIVIDUAL             39                    0   
4       INDIVIDUAL             18                    0   

   Total Collection Amount  Total Current Balance  \
0                       31                 311301   
1                       53                 182610   
2                       34                  89801   
3                       40                   9189   
4                      430                 126029   

   Total Revolving Credit Limit  Loan Status  
0                          6619            0  
1                         20885            0  
2                         26155            0  
3                         60214            0  
4                         22579            0  

[5 rows x 35 columns]

2.Display last 5 rows of the data

In [5]:
Data.tail()

ID  Loan Amount  Funded Amount  Funded Amount Investor  Term  \
70361  11093542        13154          22653            21804.279580    59   
70362   8055469        13791           9209            22065.647720    59   
70363   2142145        15299           9594            21500.619710    59   
70364  10342473        21608           5981             9952.213204    59   
70365  13613348        10517          20843            33559.365730    59   

      Batch Enrolled  Interest Rate Grade Sub Grade Home Ownership  ...  \
70361     BAT1780517      10.722019     C        E2           RENT  ...   
70362     BAT1780517      14.075933     B        E5       MORTGAGE  ...   
70363     BAT5547201      10.668933     C        B4           RENT  ...   
70364     BAT4722912       9.549941     A        A2       MORTGAGE  ...   
70365     BAT1780517       9.188369     C        C3       MORTGAGE  ...   

       Recoveries Collection Recovery Fee Collection 12 months Medical  \
70361    0.500543                1.278989                            0   
70362    3.290785                0.375653                            0   
70363    1.881125                0.502448                            0   
70364    0.470865                0.133257                            0   
70365  975.186013                0.687089                            0   

      Application Type  Last week Pay  Accounts Delinquent  \
70361       INDIVIDUAL             38                    0   
70362       INDIVIDUAL             89                    0   
70363       INDIVIDUAL            107                    0   
70364       INDIVIDUAL             14                    0   
70365       INDIVIDUAL             32                    0   

       Total Collection Amount  Total Current Balance  \
70361                       40                  63405   
70362                       29                 132166   
70363                       46                   8203   
70364                       59                 100082   
70365                      373                 293423   

       Total Revolving Credit Limit  Loan Status  
70361                          5524            0  
70362                        135755            0  
70363                         61245            0  
70364                         14706            0  
70365                          5666            0  

[5 rows x 35 columns]

3.Shape of the data

In [6]:
Data.shape

(70366, 35)

In [7]:
print('Number of rows', Data.shape[0])
print('Number of columns', Data.shape[1])

Number of rows 70366
Number of columns 35


4.Get Information of the data set

In [8]:
Data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70366 entries, 0 to 70365
Data columns (total 35 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   ID                            70366 non-null  int64  
 1   Loan Amount                   70366 non-null  int64  
 2   Funded Amount                 70366 non-null  int64  
 3   Funded Amount Investor        70366 non-null  float64
 4   Term                          70366 non-null  int64  
 5   Batch Enrolled                70366 non-null  object 
 6   Interest Rate                 70366 non-null  float64
 7   Grade                         70366 non-null  object 
 8   Sub Grade                     70366 non-null  object 
 9   Home Ownership                70366 non-null  object 
 10  Employment Duration (years)   70366 non-null  int64  
 11  Verification Status           70366 non-null  object 
 12  Payment Plan                  70366 non-null  object 
 13  L

In [9]:
# List of columns to remove for loan default prediction
columns_to_remove = ['ID', 'Funded Amount', 'Funded Amount Investor', 'Term',
                     'Batch Enrolled', 'Grade', 'Sub Grade', 'Payment Plan',
                     'Loan Title', 'Public Record', 'Collection 12 months Medical',
                     'Application Type', 'Last week Pay', 'Accounts Delinquent']

df_clean = Data.drop(columns=columns_to_remove)

In [10]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70366 entries, 0 to 70365
Data columns (total 21 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Loan Amount                   70366 non-null  int64  
 1   Interest Rate                 70366 non-null  float64
 2   Home Ownership                70366 non-null  object 
 3   Employment Duration (years)   70366 non-null  int64  
 4   Verification Status           70366 non-null  object 
 5   Debit to Income               70366 non-null  float64
 6   Delinquency - two years       70366 non-null  int64  
 7   Inquires - six months         70366 non-null  int64  
 8   Open Account                  70366 non-null  int64  
 9   Revolving Balance             70366 non-null  int64  
 10  Revolving Utilities           70366 non-null  float64
 11  Total Accounts                70366 non-null  int64  
 12  Initial List Status           70366 non-null  object 
 13  T

5.Handling catagorical columns

In [11]:
df_clean.sample(5)

Loan Amount  Interest Rate Home Ownership  Employment Duration (years)  \
21833        23574       9.438589           RENT                            4   
62092        17571      10.754002           RENT                            6   
68395         8435      15.169186       MORTGAGE                            5   
19470        15302      11.523209           RENT                           11   
39620         5505       9.807434           RENT                            9   

      Verification Status  Debit to Income  Delinquency - two years  \
21833     Source Verified        27.304690                        1   
62092     Source Verified        21.337727                        1   
68395     Source Verified        18.231239                        2   
19470            Verified        17.088345                        0   
39620            Verified        15.255327                        0   

       Inquires - six months  Open Account  Revolving Balance  ...  \
21833                      0            13               1518  ...   
62092                      0            21               1285  ...   
68395                      0             9               1859  ...   
19470                      0            14               9802  ...   
39620                      0            10               1443  ...   

       Total Accounts  Initial List Status Total Received Interest  \
21833              19                    w             1275.893934   
62092              18                    w             1079.357814   
68395               8                    w              650.224743   
19470              23                    f              472.864697   
39620              30                    w             2104.672395   

       Total Received Late Fee  Recoveries  Collection Recovery Fee  \
21833                 0.061000    2.539782                 0.758899   
62092                31.546643    4.553757                 0.309733   
68395                 0.087759    0.663706                 1.075883   
19470                 0.008908    5.018187                 1.197820   
39620                 0.048869    6.070411                 1.081010   

       Total Collection Amount  Total Current Balance  \
21833                       20                  20178   
62092                      559                 130814   
68395                        2                  84072   
19470                       48                  89593   
39620                     3208                  50216   

       Total Revolving Credit Limit  Loan Status  
21833                          1056            0  
62092                         12753            0  
68395                          3589            0  
19470                         39190            0  
39620                         19052            0  

[5 rows x 21 columns]

In [12]:
# Data encoding
df_clean['Home Ownership'] = df_clean['Home Ownership'].map({'MORTGAGE': 1, 'RENT': 2, 'OWN': 3}).astype('int')
df_clean["Verification Status"] = df_clean["Verification Status"].map({"Not Verified": 1, "Source Verified": 2, "Verified": 3}).astype('int')
df_clean["Initial List Status"] = df_clean["Initial List Status"].map({"w": 1, "f": 2}).astype('int')

print("Current shape of dataset:", df_clean.shape)

Current shape of dataset: (70366, 21)


In [13]:
df_clean.sample(5)

Loan Amount  Interest Rate  Home Ownership  \
66145        26752      10.964796               1   
51587        31634      10.517007               1   
45499        26809      15.250522               2   
40607        19505      10.317222               1   
46039        10533      13.672209               1   

       Employment Duration (years)  Verification Status  Debit to Income  \
66145                            8                    2        28.967317   
51587                            5                    3        12.042521   
45499                            7                    2        20.709742   
40607                            2                    1         6.356061   
46039                            6                    2        19.895391   

       Delinquency - two years  Inquires - six months  Open Account  \
66145                        0                      0             9   
51587                        0                      0            14   
45499                        1                      0            11   
40607                        1                      0            13   
46039                        0                      0            11   

       Revolving Balance  ...  Total Accounts  Initial List Status  \
66145               4311  ...              36                    2   
51587                 51  ...               5                    1   
45499               1289  ...              28                    2   
40607               2222  ...              39                    2   
46039               5784  ...              17                    1   

       Total Received Interest  Total Received Late Fee  Recoveries  \
66145              1220.496976                 0.079330    2.246306   
51587              4448.507041                 0.057033    4.824919   
45499                76.971741                 0.007724    8.559834   
40607               854.375094                 0.025665    4.460297   
46039               184.711734                 0.031986    6.382979   

       Collection Recovery Fee  Total Collection Amount  \
66145                 1.462682                       53   
51587                 0.711675                       21   
45499                 0.379559                       33   
40607                 0.212449                        6   
46039                 0.468160                       13   

       Total Current Balance  Total Revolving Credit Limit  Loan Status  
66145                 313883                         13247            1  
51587                 207625                         34749            0  
45499                  44518                         26538            1  
40607                  92454                          8214            0  
46039                 326453                          5471            0  

[5 rows x 21 columns]

7.Store feauture matrix in X and response in y

In [14]:
X = df_clean.drop('Loan Status', axis = 1)
y = df_clean['Loan Status']

8.Feature scaling

In [15]:
df_clean.head()

Loan Amount  Interest Rate  Home Ownership  Employment Duration (years)  \
0        10000      11.135007               1                           20   
1         3609      12.237563               2                            4   
2        28276      12.545884               1                           10   
3        11170      16.731201               1                           12   
4        16890      15.008300               1                            5   

   Verification Status  Debit to Income  Delinquency - two years  \
0                    1        16.284758                        1   
1                    2        15.412409                        0   
2                    2        28.137619                        0   
3                    2        18.043730                        1   
4                    2        17.209886                        1   

   Inquires - six months  Open Account  Revolving Balance  ...  \
0                      0            13              24246  ...   
1                      0            12                812  ...   
2                      0            14               1843  ...   
3                      0             7              13819  ...   
4                      3            13               1544  ...   

   Total Accounts  Initial List Status  Total Received Interest  \
0               7                    1              2929.646315   
1              13                    2               772.769385   
2              20                    1               863.324396   
3              12                    1               288.173196   
4              22                    1               129.239553   

   Total Received Late Fee   Recoveries  Collection Recovery Fee  \
0                 0.102055     2.498291                 0.793724   
1                 0.036181     2.377215                 0.974821   
2                18.778660     4.316277                 1.020075   
3                 0.044131     0.107020                 0.749971   
4                19.306646  1294.818751                 0.368953   

   Total Collection Amount  Total Current Balance  \
0                       31                 311301   
1                       53                 182610   
2                       34                  89801   
3                       40                   9189   
4                      430                 126029   

   Total Revolving Credit Limit  Loan Status  
0                          6619            0  
1                         20885            0  
2                         26155            0  
3                         60214            0  
4                         22579            0  

[5 rows x 21 columns]

In [16]:
cols = ['Loan Amount', 'Interest Rate', 'Employment Duration (years)', 'Debit to Income', 
        'Open Account', 'Revolving Balance', 'Total Accounts', 'Total Received Interest', 'Total Received Late Fee', 
        'Recoveries', 'Collection Recovery Fee', 'Total Collection Amount', 'Total Current Balance', 'Total Revolving Credit Limit']

In [17]:
from sklearn.preprocessing import StandardScaler
st = StandardScaler()
X[cols]=st.fit_transform(X[cols])

In [18]:
X

Loan Amount  Interest Rate  Home Ownership  \
0        -0.818578      -0.189992               1   
1        -1.582950       0.106512               2   
2         1.367254       0.189427               1   
3        -0.678645       1.314963               1   
4         0.005474       0.851632               1   
...            ...            ...             ...   
70361    -0.441356      -0.301055               2   
70362    -0.365170       0.600895               1   
70363    -0.184811      -0.315331               2   
70364     0.569753      -0.616256               1   
70365    -0.756744      -0.713492               1   

       Employment Duration (years)  Verification Status  Debit to Income  \
0                         2.194195                    1        -0.831962   
1                        -0.911469                    2        -0.935305   
2                         0.253155                    2         0.572204   
3                         0.641363                    2        -0.623583   
4                        -0.717365                    2        -0.722365   
...                            ...                  ...              ...   
70361                    -0.911469                    2         0.525458   
70362                    -0.717365                    1         0.833821   
70363                    -1.299677                    2         1.045079   
70364                     0.447259                    2         1.376640   
70365                    -1.105573                    2        -0.669540   

       Delinquency - two years  Inquires - six months  Open Account  \
0                            1                      0     -0.203901   
1                            0                      0     -0.364572   
2                            0                      0     -0.043231   
3                            1                      0     -1.167924   
4                            1                      3     -0.203901   
...                        ...                    ...           ...   
70361                        0                      0     -0.525242   
70362                        0                      0     -0.203901   
70363                        0                      0      1.884814   
70364                        0                      0     -0.685912   
70365                        0                      0      2.848837   

       Revolving Balance  Revolving Utilities  Total Accounts  \
0               2.120873            74.932551       -1.397865   
1              -0.879700            78.297186       -0.676552   
2              -0.747687             2.073040        0.164981   
3               0.785762            67.467951       -0.796770   
4              -0.785972            85.250761        0.405419   
...                  ...                  ...             ...   
70361           1.274377            78.841002       -0.556333   
70362           2.073497            49.880686        0.285200   
70363           0.660024            49.782126       -0.676552   
70364          -0.499923             4.683398       -1.277646   
70365           0.686529            54.527422        0.886295   

       Initial List Status  Total Received Interest  Total Received Late Fee  \
0                        1                 0.388844                -0.198673   
1                        2                -0.583453                -0.211236   
2                        1                -0.542632                 3.363209   
3                        1                -0.801904                -0.209720   
4                        1                -0.873550                 3.463904   
...                    ...                      ...                      ...   
70361                    2                 0.227739                -0.213020   
70362                    1                -0.850420                 4.889302   
70363                    2                -0.416925                -0.211001   
70364                  

9.Splitting the data into test and training model 

In [19]:
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score
import numpy as np

In [20]:
model_df = {}
def model_val(model,X,y):
    X_train,X_test,y_train,y_test = train_test_split(X,y,
                                                     test_size = 0.20,
                                                     random_state =42)
    
    model.fit(X_train,y_train)
    y_pred = model.predict(X_test)
    print(f"{model} acuracy is {accuracy_score(y_test,y_pred)}")

    score = cross_val_score(model,X,y,cv = 5)
    print(f"{model} Avg corss val score is {np.mean(score)}")
    model_df[model] = round(np.mean(score)*100,2)

10.Logistic regression

In [21]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()
model_val(model,X,y)

LogisticRegression() acuracy is 0.9135284922552224
LogisticRegression() Avg corss val score is 0.9113065976721939


12.Decision tree classifier

In [22]:
from sklearn.tree import DecisionTreeClassifier
model = DecisionTreeClassifier()
model_val(model,X,y)

DecisionTreeClassifier() acuracy is 0.8222964331391218
DecisionTreeClassifier() Avg corss val score is 0.8235511424770943


13.Random forest classifier

In [23]:
from sklearn.ensemble import RandomForestClassifier
model = RandomForestClassifier()
model_val(model,X,y)


RandomForestClassifier() acuracy is 0.9135284922552224
RandomForestClassifier() Avg corss val score is 0.9112923870710855


14.Gradient boosting classifier

In [24]:
from sklearn.ensemble import GradientBoostingClassifier
model = GradientBoostingClassifier()
model_val(model,X,y)

GradientBoostingClassifier() acuracy is 0.9133863862441381
GradientBoostingClassifier() Avg corss val score is 0.9112355406275411


15.Hyperparameter tunning

In [25]:
from sklearn.model_selection import RandomizedSearchCV

Logistic regression

In [26]:
log_reg_grid = {"C":np.logspace(-4,4,20),
                "solver":['liblinear']}

In [27]:
rs_log_reg = RandomizedSearchCV(LogisticRegression(),
                   param_distributions = log_reg_grid,
                   n_iter = 20, cv = 5, verbose = True)

In [28]:
rs_log_reg.fit(X,y)

Fitting 5 folds for each of 20 candidates, totalling 100 fits


RandomizedSearchCV(cv=5, estimator=LogisticRegression(), n_iter=20,
                   param_distributions={'C': array([1.00000000e-04, 2.63665090e-04, 6.95192796e-04, 1.83298071e-03,
       4.83293024e-03, 1.27427499e-02, 3.35981829e-02, 8.85866790e-02,
       2.33572147e-01, 6.15848211e-01, 1.62377674e+00, 4.28133240e+00,
       1.12883789e+01, 2.97635144e+01, 7.84759970e+01, 2.06913808e+02,
       5.45559478e+02, 1.43844989e+03, 3.79269019e+03, 1.00000000e+04]),
                                        'solver': ['liblinear']},
                   verbose=True)

In [29]:
rs_log_reg.best_score_

0.9113065976721939

In [30]:
rs_log_reg.best_params_

{'solver': 'liblinear', 'C': 0.0001}

Random forest classifier

In [31]:
RandomForestClassifier()

RandomForestClassifier()

In [32]:
rf_grid = {'n_estimators': np.arange(10,100,10),
'max_features' : ['auto', 'sqrt'],
'max_depth' : [None,10,20],
'min_samples_split' : [2,100],
'min_samples_leaf' : [1,5] }

In [33]:
rs_rf = RandomizedSearchCV(RandomForestClassifier(),
                            param_distributions=rf_grid,
                            cv=5,
                            n_iter=10,
                            n_jobs=-1,
                            verbose=True)

In [34]:
rs_rf.fit(X,y)

Fitting 5 folds for each of 10 candidates, totalling 50 fits


RandomizedSearchCV(cv=5, estimator=RandomForestClassifier(), n_jobs=-1,
                   param_distributions={'max_depth': [None, 10, 20],
                                        'max_features': ['auto', 'sqrt'],
                                        'min_samples_leaf': [1, 5],
                                        'min_samples_split': [2, 100],
                                        'n_estimators': array([10, 20, 30, 40, 50, 60, 70, 80, 90])},
                   verbose=True)

In [35]:
rs_rf.best_score_

0.9113065976721939

In [36]:
rs_rf.best_params_

{'n_estimators': 60,
 'min_samples_split': 100,
 'min_samples_leaf': 5,
 'max_features': 'sqrt',
 'max_depth': 20}

Logistic regression score before hyperparameters tunning = 0.9113065976721939

Logistic regression score after hyperparameters tunning = 0.9113065976721939

________

Random forest classifier score before hyperparameters tunning = 0.9112923870710855

Random forest classifier score after hyperparameters tunning = 0.9113065976721939



Save the model 

saving the model of random forest classifier which can be changed as per needs

In [37]:
X = df_clean.drop('Loan Status', axis = 1)
y = df_clean['Loan Status']

In [38]:
rf = RandomForestClassifier(n_estimators = 20,
 min_samples_split = 2,
 min_samples_leaf = 5,
 max_features = 'auto',
 max_depth = None)

In [39]:
rf.fit(X,y)

c:\Users\Ryzen\anaconda3\envs\dev\lib\site-packages\sklearn\ensemble\_forest.py:424: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(


RandomForestClassifier(max_features='auto', min_samples_leaf=5, n_estimators=20)

In [40]:
import joblib

In [41]:
joblib.dump(rf, 'loan_status_prediction')

['loan_status_prediction']